### PR TITLE
[Prefix] List reserved namespaces in the manage packages page

### PR DESCRIPTION
**Pull request by shishirx34**
The view looks like this:

image

### CODE REVIEW

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+﻿@model ReservedNamespaceViewModel
+@{
+    if (!Model.ReservedNamespaces.Any())
+    {
+        return;
+    }
+}
+
+<div class="row user-package-list">
+    <div class="col-md-12">
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: `col-xs-12`

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        <h2>
+            <a href="#" role="button" data-toggle="collapse" data-target="#reservednamespaces"
+               aria-expanded="true" aria-controls="reservednamespaces" id="show-reservednamespaces">
+                <i class="ms-Icon ms-Icon--ChevronDown" aria-hidden="true"></i>
+                <span>My Reserved Package Ids</span>
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: `IDs` not `Ids`

File: src/NuGetGallery/ViewModels/ReservedNamespaceViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+namespace NuGetGallery
+{
+    public class ReservedNamespaceViewModel
+    {
+        public IEnumerable<ReservedNamespaceListItemViewModel> ReservedNamespaces { get; }
+
+        public ReservedNamespaceViewModel(ICollection<ReservedNamespace> reservedNamespacesList)
+        {
+            ReservedNamespaces = reservedNamespacesList
+                .ToList()
```
Comment (loic-sharma) [inline src/NuGetGallery/ViewModels/ReservedNamespaceViewModel.cs]: The `ToList` should be after the `Select`, right?

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                                <td class="align-middle">
+                                    <a target="_blank" href="@Url.Search(reservedNamespace.Value)">@reservedNamespace.GetPattern()</a>
+                                </td>
+                                <td class="align-middle">
+                                    @foreach (var owner in reservedNamespace.Owners)
+                                    {
+                                        <a target="_blank" href="@Url.User(owner.Username)">@owner.Username</a>
+                                    }
+                                </td>
+                                @if(reservedNamespace.IsPublic)
```
Comment (loic-sharma) [inline src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: Add a space between `@if` and `(`

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml (hunk 1 above)
Comment (shishirx34) [reply src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: looks like `col-md-12` behaves correctly across different dimensions. Will leave as is.

File: src/NuGetGallery/ViewModels/ReservedNamespaceViewModel.cs (hunk 3 above)
Comment (shishirx34) [reply src/NuGetGallery/ViewModels/ReservedNamespaceViewModel.cs]: Don't need it actually, leftover from previous change. removed.

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml (hunk 4 above)
Comment (shishirx34) [reply src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: formatted the document.

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml (hunk 2 above)
Comment (scottbommarito) [inline src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: IMO I'd prefer
```
My Reserved Packages
```
or even
```
My Reserved Package Namespaces
```

I don't think we explicitly say `IDs` prominently in the UI anywhere else.

File: src/NuGetGallery/ViewModels/ReservedNamespaceViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGetGallery
+{
+    public class ReservedNamespaceViewModel
```
Comment (scottbommarito) [inline src/NuGetGallery/ViewModels/ReservedNamespaceViewModel.cs]: To follow the pattern of all of our other list view models this should be `ReservedNamespaceListViewModel`

File: src/NuGetGallery/Views/Users/Packages.cshtml [hunk 6]
```diff
@@ -20,6 +20,10 @@
     }
 
     @{
+        @Html.Partial("_ReservedNamespacesList", Model.ReservedNamespaces)
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Users/Packages.cshtml]: nit: No need to wrap this in `@{ ... }`

File: src/NuGetGallery/ViewModels/ReservedNamespaceListItemViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        public bool IsPrefix { get; }
+
+        public IEnumerable<User> Owners { get; }
+
+        public ReservedNamespaceListItemViewModel(ReservedNamespace reservedNamespace)
+        {
+            Value = reservedNamespace.Value;
+            IsPublic = reservedNamespace.IsSharedNamespace;
+            IsPrefix = reservedNamespace.IsPrefix;
+            Owners = reservedNamespace.Owners.Select(owner => owner);
```
Comment (scottbommarito) [inline src/NuGetGallery/ViewModels/ReservedNamespaceListItemViewModel.cs]: Why do you need this `Select`?

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml [hunk 8]
```diff
@@ -0,0 +1,63 @@
+﻿@model ReservedNamespaceViewModel
+@{
+    if (!Model.ReservedNamespaces.Any())
+    {
+        return;
+    }
+}
+
+<div class="row user-package-list">
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: Given that this is the second clone of `_UserPackagesList.cshtml` that we've implemented, can we create some sort of ViewHelpers function that can create a list like this? Every time one page changes, we have to change every other page.

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                            <th>Owners</th>
+                            <th>Upload Restrictions</th>
+                        </tr>
+                    </thead>
+                    <tbody>
+                        @foreach (var reservedNamespace in @Model.ReservedNamespaces)
+                        {
+                            <tr class="manage-package-listing">
+                                <td class="align-middle hidden-xs">
+                                    <img class="reserved-indicator-icon img-responsive"
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: nit: Do we have this same chunk of code everywhere we embed the reserved package icon? Is there some way we could put it in a shared place?

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 10]
```diff
@@ -161,11 +161,13 @@ public virtual ActionResult Packages()
             var outgoing = _packageOwnerRequestService.GetPackageOwnershipRequests(requestingOwner: user);
 
             var ownerRequests = new OwnerRequestsViewModel(incoming, outgoing, user, _packageService);
+            var reservedPrefixes = new ReservedNamespaceViewModel(user.ReservedNamespaces);
```
Comment (scottbommarito) [inline src/NuGetGallery/Controllers/UsersController.cs]: Are we concerned we're adding too many lists to `ManagePackages`? It has the user's packages, their owner requests, and now their reserved namespaces. It's ok now, but it just seems like we're dumping tons of content there.

Comment (shishirx34) [reply src/NuGetGallery/Controllers/UsersController.cs]: There isn't a better place for this, I don't think this warrants another page either. Also, given that this  would show up only if you have a reserved namespace and same goes with owner requests I think its fine.

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml (hunk 2 above)
Comment (shishirx34) [reply src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: This is as per @diverdan92 's spec. I think IDs clearly articulates the message here.

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml (hunk 9 above)
Comment (shishirx34) [reply src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: I don't think we should refactor the view elements much, this isn't exactly the same piece of code, the fallback urls are different, its easier to change the view elements that are aggregated together.

File: src/NuGetGallery/ViewModels/ReservedNamespaceListItemViewModel.cs (hunk 7 above)
Comment (shishirx34) [reply src/NuGetGallery/ViewModels/ReservedNamespaceListItemViewModel.cs]: don't need it, reminiscent of some view model changes I was making. removed it.

File: src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml (hunk 8 above)
Comment (shishirx34) [reply src/NuGetGallery/Views/Users/_ReservedNamespacesList.cshtml]: Perhaps we can, but I wouldn't want to make it for this PR given that its a much bigger change for refactoring and affecting other components for thorough testing.

**Pull request merged**